keras_nlp/models/falcon/falcon_transformer_decoder.py

```diff
@@ -0,0 +1,145 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from keras_nlp.backend import keras
+
+
+class FalconTransformerDecoder(keras.layers.Layer):
+    # TODO: set default values based on Falcon's config.
+    def __init__(
+        self,
+        num_attention_heads,
+        layer_norm_epsilon=1e-5,
+        attention_dropout=0,
+        feedforward_dropout=0,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.num_attention_heads = num_attention_heads
+        self.layer_norm_epsilon = layer_norm_epsilon
+        self.attention_dropout = attention_dropout
+        self.feedforward_dropout = feedforward_dropout
+
+    def build(self, decoder_sequence_shape):
+        self.hidden_dim = decoder_sequence_shape[-1]
+        self._input_layernorm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon
+        )
+        self._input_layernorm.build(decoder_sequence_shape)
+
+        # Attention layers.
+        # MHA for 1b model.
+        self.key_dim = self.hidden_dim // self.num_attention_heads
+        self._attention_layer = keras.layers.MultiHeadAttention(
+            num_heads=self.num_attention_heads,
+            key_dim=self.key_dim,
+        )
+        self._attention_layer.build(
+            decoder_sequence_shape,
+        )
+
+        self._attention_dropout = keras.layers.Dropout(
+            rate=self.attention_dropout,
+            name="attention_dropout",
+        )
+
+        self._post_attention_layernorm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon
+        )
+        self._post_attention_layernorm.build(decoder_sequence_shape)
+
+        # Feedforward layers.
+        self._dense_h_to_4h = keras.layers.Dense(
+            units=4 * self.hidden_dim,
+            input_dim=self.hidden_dim,
+            activation=keras.activations.gelu,
+            use_bias=False,
+            name="dense_h_to_4h",
+        )
+        self._dense_h_to_4h.build(decoder_sequence_shape)
+
+        self._dense_4h_to_h = keras.layers.Dense(
+            units=self.hidden_dim,
+            input_dim=4 * self.hidden_dim,
+            use_bias=False,
+            name="dense_4h_to_h",
+        )
+        self._dense_4h_to_h.build(
+            (
+                decoder_sequence_shape[0],
+                decoder_sequence_shape[1],
+                4 * decoder_sequence_shape[2],
+            )
+        )
+
+        self._feedforward_dropout = keras.layers.Dropout(
+            rate=self.feedforward_dropout,
+            name="feedforward_dropout",
+        )
+
+        self.built = True
+
+    def call(
+        self,
+        decoder_sequence,
+        # attention_cache=None,
+        # attention_cache_update_index=None,
+        training=None,
+    ):
+        # TODO: do we need self._compute_self_attention_mask?
```

Comment:
You will definitely want something that adds a causal attention mask here, otherwise we won't match reference output for any sequence with length > 1. `_compute_self_attention_mask` is probably a good place to start.

Reply:
Thanks for explaining this! I added `_compute_self_attention_mask` similar to the `llama` implementation for now. I'll make adjustment if needed.